CMakeLists.txt

```diff
@@ -161,31 +165,48 @@ set(CMAKE_MESSAGE_LOG_LEVEL NOTICE)
 
 # Add third party subdirectories and define options appropriate to run their cmakes.
 set(pybind11_FIND_QUIETLY ON)
-add_subdirectory(third_party/onnx)
-add_subdirectory(third_party/pybind11)
-add_subdirectory(third_party/rapidcheck)
+
+if (ONNX_MLIR_USE_SYSTEM_ONNX)
+  find_package(ONNX REQUIRED)
```

Comment:
onnx-mlir requires a specific version of onnx, which is the one included under third_party. So there needs some way to extract that version number and put it here.

Reply:
@gongsu832 ,

* The ```find_package(ONNX REQUIRED)``` is filling the ```ONNIX_VERSION``` cmake var already.
* It is also [picked up and used here](https://github.com/onnx/onnx-mlir/pull/2698/files#diff-9c32b95937462105ee738c4d993a32c9669dfefb9b73742ca8021e2db41cdd53R89-R91) in the case when a local system library is found.

See a runtime sample:
```
$ rpm -q onnx-devel
onnx-devel-1.16.0-20240123.1.git3a39b0b8.fc40.x86_64

$ cmake  {..} -DONNX_MLIR_USE_SYSTEM_ONNX=ON {...}
{...}
-- Accelerators             : NONE
-- OpenMP support           : OFF
-- Found Git: /usr/bin/git (found version "2.43.0") 
-- Onnx version             : 1.16.0
{...}
```

* I did let it by default ```OFF``` these knobs (so default is to pick up ```third_party```), not enforced a specific version as ```REQUIRED```, one who want to use these is assuming to know what he is doing.

Reply:
onnx-mlir currently does NOT support onnx 1.16.0. It only works with a specific version of onnx, currently 1.15.0. That's why we always install onnx from third_party/onnx. If you want to use a system installed onnx, it must match the version found in third_party/onnx.

Reply:
> @gongsu832 ,
> 
> * The `find_package(ONNX REQUIRED)` is filling the `ONNIX_VERSION` cmake var already.
> * It is also [picked up and used here](https://github.com/onnx/onnx-mlir/pull/2698/files#diff-9c32b95937462105ee738c4d993a32c9669dfefb9b73742ca8021e2db41cdd53R89-R91) in the case when a local system library is found.
> 
> See a runtime sample:
> 
> ```
> $ rpm -q onnx-devel
> onnx-devel-1.16.0-20240123.1.git3a39b0b8.fc40.x86_64
> 
> $ cmake  {..} -DONNX_MLIR_USE_SYSTEM_ONNX=ON {...}
> {...}
> -- Accelerators             : NONE
> -- OpenMP support           : OFF
> -- Found Git: /usr/bin/git (found version "2.43.0") 
> -- Onnx version             : 1.16.0
> {...}
> ```
> 
> * I did let it by default `OFF` these knobs (so default is to pick up `third_party`), not enforced a specific version as `REQUIRED`, one who want to use these is assuming to know what he is doing.

@gongsu832 ,

To better reflect the origin of ONNX I [added explicit message](https://github.com/onnx/onnx-mlir/pull/2698/files#diff-9c32b95937462105ee738c4d993a32c9669dfefb9b73742ca8021e2db41cdd53R94-R98) for the case of system found library.
```
$ cmake  {..} -DONNX_MLIR_USE_SYSTEM_ONNX=ON {...}
{...}
-- Onnx version             : 1.16.0 (system)

$ cmake  {..} -DONNX_MLIR_USE_SYSTEM_ONNX=OFF {...}
-- Onnx version             : 1.16.0
```

Reply:
@gongsu832 

> onnx-mlir currently does NOT support onnx 1.16.0. It only works with a specific version of onnx, currently 1.15.0. That's why we always install onnx from third_party/onnx. If you want to use a system installed onnx, it must match the version found in third_party/onnx.

Ok, it is [enforced now](https://github.com/onnx/onnx-mlir/pull/2698/files#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR170) to specific version.
```
$ cmake  {..} -DONNX_MLIR_USE_SYSTEM_ONNX=ON {...}
{...}

CMake Error at CMakeLists.txt:173 (find_package):
  Could not find a configuration file for package "ONNX" that exactly matches
  requested version "1.15.0".

  The following configuration files were considered but not accepted:

    /usr/lib64/cmake/ONNX/ONNXConfig.cmake, version: 1.16.0
    /lib64/cmake/ONNX/ONNXConfig.cmake, version: 1.16.0

```
However the idea of this permissive know was that once user enable it he *really knows* what is doing.

Reply:
@cbalint13 As I said above, onnx-mlir only works with a specific version of onnx, currently 1.15.0. In fact, I think the presence of third_party/onnx source repo is mandatory (@chentong319 can confirm yes or no) in order to generate the operator translation table. That's precisely why we always install onnx from third_party/onnx instead of having to deal with incorrect version of system installed onnx.

Reply:
@gongsu832 , @chentong319 

> As I said above, onnx-mlir only works with a specific version of onnx, currently 1.15.0. 

* Let's have it enforced, in case of system lib too, just like it is now: 
https://github.com/onnx/onnx-mlir/pull/2698/files#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR170

> In fact, I think the presence of third_party/onnx source repo is mandatory

* Works fine *without*, I can confirm from my side, one needed workaround for this is here:
 https://github.com/onnx/onnx-mlir/pull/2698/files#diff-9c32b95937462105ee738c4d993a32c9669dfefb9b73742ca8021e2db41cdd53R89-R91

Reply:
When you say it "works fine _without_", I assume you mean that cmake runs fine. But it takes more to upgrade onnx to a new version for onnx-mlir. For example, you also need to run `make OMONNXOpsIncTranslation` from the `build/utils` directory to generate the operator translation table. I think running `make OMONNXOpsIncTranslation` requires the presence of third_party/onnx source repo. Of course, once the table has been generated, we may not need the third_party/onnx source repo anymore.

Reply:
> When you say it "works fine _without_", I assume you mean that cmake runs fine. But it takes more to upgrade onnx to a new version for onnx-mlir. 

* Works fine *without* the ```/third_party/onnx``` (absent at all), having ```1.15.0 EXACT REQUIRED``` system version.
* By *works* I mean: configures + compiles + tests

>For example, you also need to run `make OMONNXOpsIncTranslation` from the `build/utils` directory to >generate the operator translation table. I think running `make OMONNXOpsIncTranslation` requires the >presence of third_party/onnx source repo. 

* All include headers and other ONNX things are well picked up in the context of system libraries, find_package(ONNX ...) provides all required paths to the onnx requirements.